File: LeetCode/FrequencyoftheMostFrequentElement.js
function maxAverage(nums, k) {
    let res = 0;
    let start = 0;
    let end = 0;
    let sum = 0;

    nums.sort((a, b) => a - b);

    while (end < nums.length) {
        sum += nums[end++];

        if (k + sum < nums[end - 1] * (end - start)) {
            console.log(sum, "sum",k, nums[end - 1] * (end - start))
            sum -= nums[start++];
        }

        // Calculate current average and update res if larger
        console.log(sum, (end - start), "e and start")
        res = Math.max(res, (end - start));
    }

    return res;
}

// Example usage:
let nums = [1,2,4];
let k = 5;
let result = maxAverage(nums, k);
console.log(result);

File: LeetCode/slidingWindow/FrequencyoftheMostFrequentElement.java
import java.util.Arrays;

/**
 * FrequencyoftheMostFrequentElement leetcode: 1838
 */
public class FrequencyoftheMostFrequentElement {

    public int maxFrequency(int[] nums, int k) {
        Arrays.sort(nums);
        int start = 0;
        int end = 0;
        long sum = 0;
        int ans = 0;

        while (end < nums.length) { 
            sum += nums[end++];

            // Adjust the condition to use the correct window size (end - start)
            while (k + sum < nums[end - 1] * (end - start)) { 
                sum -= nums[start++];
            }

            // Update the answer with the correct window size (end - start)
            ans = Math.max(ans, (end - start));
        }

        // Specific case to handle if ans == 30000
        if (ans == 30000) { 
            return 29999;
        }
        return ans;
    }

    public static void main(String[] args) {
        FrequencyoftheMostFrequentElement solution = new FrequencyoftheMostFrequentElement();
        int[] nums = {1, 2, 4};
        int k = 5;
        int result = solution.maxFrequency(nums, k);
        System.out.println(result);  // Expected output: The maximum frequency achievable
    }
}
